### DISCUSSION TITLE
Welcome to grblHAL-teensy-4.x Discussions!

### DISCUSSION THREAD

**Discussion by phil-barrett** [phil-barrett/grblHAL-teensy-4.x#25] [Announcements]
## 👋 Welcome!
  We’re using Discussions as a place to connect with other members of our community. We hope that you:
  * Ask questions you’re wondering about.
  * Seek help on building your grblHAL based machine.
  * Share ideas.
  * Give suggestions for future grblHAL based products.
  * Engage with other community members.
  * Welcome others and are open-minded. Remember that this is a community we
  build together 💪.

  To get started, comment below with an introduction of yourself and tell us about what you do with this community.

**Comment by karoria** [comment]
Hi Phil, this is a good initiative. I am Ravi from India. I am building my controller from scratch with grblHAL-Teensy4.1. This is because I need some customizations for my machine. I am also writing a basic web app using html/css/vanilla js, using websockets via ethernet to make it more "industrial". This app will be hosted in sd card or flash memory of controller. So basically you need a pc or sbc for a browser which just open my web app in full screen, then the UI will accept commands and communicate to controller via websockets. I am almost more than half in this way. Let me tell you that I am a man of Mechanical. For software, I have just beginner to intermediate level skills in JS only. For firmware, I am not even having basic skills for C. I am learning that as time permits. 😊

**Reply by phil-barrett** [reply]
Welcome, Ravi. Keep us informed of your progress.

**Reply by karoria** [reply]
Sure Phil. I may need your occassional help too. Your instructions on magjack ethernet done me great help and I got it working. I will see if I can make my app universal for grblHAL, if so, I will make it opensource. Actually, right now, it has so many machine specific code. It is optimized for 6 to 10 inch touch displays. So better operated with a mobile phone, raspberry pi with its 7 inch native display, tablets, etc. It has good amount of keyboard shortcuts for almost all functions.

**Comment by stevempotter** [comment]
Thanks, Phil, for turning on the Discussions! I predict it will be a busy forum.
Regarding using proximity sensors for endstops, I have implemented the use of Hall Effect Switches as my endstops and they work great. They are much cheaper and simpler to use than the metal-sensing endstops (just need a pull-up resistor on the plus line). They will never wear out or get clogged with sawdust. See them on your favourite supplier: [part AH3391Q-P-B ](https://www2.mouser.com/ProductDetail/Diodes-Incorporated/AH3391Q-P-B?qs=%2Fha2pyFadujpqSQV2kX0tkfFydYtewTiQS0BpC%252BDdcd%2FDLdZ5%2FpZlA==)
They have the advantage that they will not be mistriggered by a random metal object coming near them, it has to be the south pole of a magnet to trigger them. I chose the least sensitive ones to be less influenced by the Earth's magnetic field. (My CNC is portable and changes orientation at times).

**Reply by phil-barrett** [reply]
Yes, that is a good HE sensor.  I agree with magnet vs metal prox (all the prox sensor use HE but supply the magnet and the metal target changes the field but I digress). I have triggered the Pepperl-Fuchs sensors on the Pro4848 with my belt buckle in the middle of a job.  Ugh. 

There are plans out there to build them.  Uses an LED near the sensor so you see it working.  I always wondered if there was a DIY product.  Only a tiny bit of the cost is in the electronics.

**Comment by molove** [comment]
Thanks Phil for setting up Discussions. 

I'm Piers from London, I'm a hobbist, I've owned a 3018 for a while and made a load of modifications to it to make it more rigid and a more powerful spindle. Whilst it was an improvement, it wasn't great so when I discovered Phil's breakout board I decided to try to build a better CNC from scratch. 

Being a noob I'm bound to have a load of questions about the control side and setting up GRBLHal.

**Comment by karoria** [comment]
Hi everyone,
I am facing a strange issue. I have my own custom PCB running grblHAL with Teensy4.1. I use websockets as main communication to my host pc. The webui code is kept in sd card in a directory named "www". That runs great normally unless I connect my limit switches to the board. When limits are connected with board and is powered on, the ethernet connection hangs! I can  confirm that, if limit switches are connected after board boots up, it works but little sluggishly. I use 24v industrial inductive proximity with optoisolation. For your kind consideration, I am trying to attach my circuit diagram. I use active high type switches. When they are triggered, the pin comes high but at the controller side of opto is reverse logic, i.e. active low. Any clues, what to check?![Screenshot_20211211-224818_Drive.jpg](https://user-images.githubusercontent.com/45288223/145685604-2693243b-d5e9-4017-913f-c97879cc6064.jpg)

**Reply by phil-barrett** [reply]
I don't see anything obvious though can only see part of your schematic.  How are you powering the Teensy?

**Reply by karoria** [reply]
It is a 24v to 5v 2A buck converter. All my external devices (inputs and outputs) are 24v compatible. I could not take a screenshot with sufficient resolution, that is why I thought to zoom at the portion in doubt. I am trying now for mostly include all.![Screenshot_20211212-073238_Drive.jpg](https://user-images.githubusercontent.com/45288223/145697652-27a9bef8-e13d-48a7-89ed-a5f21a836731.jpg)
Buck converter circuit is just above title block. C24 and C25 are 10uF.

**Reply by karoria** [reply]
> It is a 24v to 5v 2A buck converter. All my external devices (inputs and outputs) are 24v compatible. I could not take a screenshot with sufficient resolution, that is why I thought to zoom at the portion in doubt. I am trying now for mostly include all.![Screenshot_20211212-073238_Drive.jpg](https://user-images.githubusercontent.com/45288223/145697652-27a9bef8-e13d-48a7-89ed-a5f21a836731.jpg)
> Buck converter circuit is just above title block. C24 and C25 are 10uF.

Sorry, I am unable to attach pdf here. I think, they don't like it. :)

**Reply by phil-barrett** [reply]
Nothing looks obvious. Have you tried to see if the teensy is rebooting when you plug in?  I'd also look to see if the power supply is getting pulled low.

**Reply by karoria** [reply]
Teensy is not rebooting. Supply is also ok. I suspect EMI. Will try some improvements tomorrow.

**Reply by phil-barrett** [reply]
EMI sounds unlikely.  You have belts and suspenders...  What would the source be? Do you have motors or VFD running during your tests?

You say it runs sluggishly - I'd try to figure out why that is the case.  You can output diagnostic info on the serial port.  I do not recall the interface but I am sure Terje would tell you.

And, are you running a "bare essentials" grblHAL build?  Just ethernet and no other plugins.

**Reply by phil-barrett** [reply]
BTW, I am still suspicious of your power supply.  Try running the 5V/3.3V system from USB power. (ie completely isolate from the 24VDC source)

**Reply by karoria** [reply]
Will try that today and see.

**Reply by karoria** [reply]
One more observation. Now I don't even get the "sluggish" connection. Weired. I have not made any change in my setup. So, now I get connected when my limit switches are removed. Whether at boot time or after booting. My doubt of EMI is due to 24v and Ov power lines of proximity. They should be inviting noise rather than signal line which is opto isolated. Still, I will see if a separate usb power instead of 24v can work. But in that case power to proximity cuts, so we may not come to any concrete observation.

**Reply by karoria** [reply]
> BTW, I am still suspicious of your power supply.  Try running the 5V/3.3V system from USB power. (ie completely isolate from the 24VDC source)

You were right Phil. It ran successfully on 5v usb power, then I came to know that my 24v smps was not getting proper ground (earth). Solved now. Thanks for your help.
